Add `<HtmlPanel>` component

Asha20 opened the pull request:
This PR adds the `<HtmlPanel>` component.

I've also updated the example app to include an HTML panel so you can see it in action.

Asha20 committed:
feat: Add `<HtmlPanel>` component

Asha20 committed:
chore: Update `package.json`

Asha20 committed:
chore: Fix import

eteeselink commented on README.md:
```diff
@@ -274,13 +274,28 @@ Accepted props:
   loading
 - `chatboxRef` (resp. `inboxRef`, `popupRef`) - Pass a ref (created with `useRef`) and it'll be set to the vanilla JS [Chatbox](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Chatbox/) (resp. [Inbox](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Inbox/), [Popup](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Popup/)) instance. See [above](#using-refs) for an example.
 - All [Talk.ChatboxOptions](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Session/#ChatboxOptions)
+- `children?: ReactNode` - Optional. You can provide an `<HtmlPanel>` component as a child to use [HTML Panels](https://talkjs.com/docs/Features/Customizations/HTML_Panels/).
```
I think this should explicitly say that HtmlPanels can be the only direct children

eteeselink commented on README.md:
```diff
@@ -274,13 +274,28 @@ Accepted props:
+
+Accepted props:
+
+- `url: string` - The URL you want to load inside the HTML panel. The URL can be absolute or relative. We recommend using same origin pages to have better control of the page. Learn more about HTML Panels and same origin pages [here](https://talkjs.com/docs/Features/Customizations/HTML_Panels/)
```
This could be more precise, “to have better control” is a bit meaningless and also kinda wrong. It _must_ be the same origin if they want to render children into it. If they don’t want to render their contents right here from React then the origin doesn’t matter. (Btw make sure you test both cases, ie that if there’s no children passed, and the panel has a different origin, there are no errors).

eteeselink commented on README.md:
```diff
@@ -274,13 +274,28 @@ Accepted props:
+
+- `height?: number` - Optional. The panel height in pixels. Defaults to `100px`.
+
+- `show?: boolean` - Optional. Sets the visibility of the panel. Defaults to `true`.
```
Worth explaining how this is different vs conditionally rendering the HtmlPanel element

eteeselink reviewed (commented):
First half of my review, second half coming up

eteeselink commented on lib/HtmlPanel.tsx:
```diff
@@ -0,0 +1,94 @@
+
+    return () => {
+      if (state.type === "loaded") {
+        state.panel.destroy();
```
shouldn't we always destroy it, regardless of loading state? else if customer code creates it and then quickly removes it, the chatbox will create it anyway.

eteeselink replied:
(maybe just put the `Promise<HtmlPanel>` in the state, and do `state.panelPromise.then(panel => panel.destroy())`. i don't think there's a downside to already-resolved panels being destroyed in the next microtask instead of right there in the cleanup callback, right?

eteeselink commented on lib/HtmlPanel.tsx:
```diff
@@ -0,0 +1,94 @@
+        state.panel.hide();
+      }
+    }
+  }, [state, show]);
```
shouldn't this be `state.type` instead of the entire state? (same for `height`)

eteeselink commented on lib/HtmlPanel.tsx:
```diff
@@ -0,0 +1,94 @@
+
+  useEffect(() => {
+    if (state.type === "loaded") {
+      if (show) {
+        state.panel.show();
+      } else {
+        state.panel.hide();
+      }
```
i love simple & boring code like this

eteeselink commented on lib/HtmlPanel.tsx:
```diff
@@ -0,0 +1,94 @@
+  }, [state, show]);
+
+  return (
+    <>
+      {state.type === "loaded" &&
+        createPortal(children, state.panel.window.document.body)}
+    </>
```
magic, isn't it? :D

btw i think this needs some "if no children given, then no portal" type logic, so we dont get errors when people mount non-same-origin html panels (eg because they're SSR'ing something from somewhere else)

eteeselink commented on lib/HtmlPanel.tsx:
```diff
@@ -0,0 +1,94 @@
+      if (state.type !== "none" || !box) return;
+
+      setState({ type: "loading" });
+      const panel = await box.createHtmlPanel({
+        url,
+        conversation: conversationId,
+        height,
+        show,
+      });
```
meta comment, I just realized that a set of `<HtmlPanel>` components in a react app is semantically equivalent to the `htmlPanelOptions` we pass from the JSSDK in the frontend.

this is actually super nuts. we internally have a declarative data structure, we built an imperative interface around that to mutate that data structure (createHtmlPanel, destroy, show, etc), and now here we're building the same thing in the opposite direction!

i don't think we need to do anything rn with this observation but i'm thinking that maybe one day we should deprecate all `create*` and `set*` methods entirely in favour of some vanilla JS, react-y declarative "spec your chatbox here" interface. would make the entire react sdk a near no-op, also.

eteeselink reviewed (changes requested):
love it. got lots of nits tho cause i'm like that

eteeselink commented on example/panel.html:
```diff
@@ -0,0 +1,19 @@
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
```
This is a bit meaningless, no? Maybe just keep the title out

eteeselink commented on example/panel.html:
```diff
@@ -0,0 +1,19 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
+    <style>
+      body {
```
Maybe let's just give `body, html` a `padding: 0` for good measure?  (and eg a margin on the button). i expect anybody will want to do that, so the html panel behaves pretty much like a div.

Asha20 replied:
Not sure what you mean here; the buttons already have `margin: 0.6rem auto`

eteeselink commented on example/panel.html:
```diff
@@ -0,0 +1,19 @@
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
+    <style>
```
How about we prepend something here like
```html
<link rel="stylesheet" href="./your-styles.css">
<!-- Put your app's CSS here. For instance, if your bundler generates a CSS file from all component styles, load it here as well and your components will be styled correctly inside the HTML Panel -->
```

eteeselink commented on lib/MountedBox.tsx:
```diff
@@ -11,31 +11,37 @@ interface Props {
       )}
 
       <div ref={ref} style={style} className={className} />
+      {children}
```
maybe we should error out if there are children that aren't HTML Panels? Theoretically you could put anything in there, it's just a bit meaningless though, as it'll be identical to putting the children after the chatbox. If we error out then at least we don't accidentally allow something that breaks backward compat in some unspecified future.

eteeselink reviewed (changes requested):
Love it. I think I reviewed everything now!

Asha20 committed:
wip

Asha20 committed:
fix: Panels not being destroyed properly on component unmount

Asha20 committed:
feat: Validate children of UIBox components

Asha20 committed:
refactor: Apply review comments

Asha20 committed:
fix: `Inbox` and `Popup` re-rendering needlessly

Previously we'd incorrectly pass their `children` into `useUIBox` which
would cause the UIBoxes to be destroyed and re-initialized over and over.

lechinoix commented:
Hello there ! Thank you so much for your Pull Request, I'll move to your fork as I definitely need it :) Might be nice to merge it at some point

lechinoix commented:
@eteeselink Arf actually I cannot manage to use this branch directly, no types etc... Could you release it on a canary version or whatsoever ? It would be of great help :)

Asha20 committed:
Merge branch 'master' into feat/htmlpanel

Asha20 committed:
chore: v0.1.6-beta.0

Asha20 commented:
The experimental `<HtmlPanel>` component is now available by installing the `@talkjs/react` library with the `next` tag:

```
npm install @talkjs/react@next
```

Asha20 committed:
chore: v0.1.7-beta.0

Merge branch 'master' into feat/htmlpanel

frederikcreemers committed:
Merge branch 'master' into feat/htmlpanel

Asha20 committed:
chore: v0.1.9-beta.0

Asha20 committed:
fix(htmlpanel) Properly reeinitialize session when user ID changes

Asha20 committed:
fix(htmlpanel) Reinitialize session when auth params change

Asha20 committed:
Merge pull request #13 from talkjs/fix/htmlpanel_crash

(Fix) HtmlPanel crash bug

Asha20 committed:
chore: v0.1.9-beta.1
